### DISCUSSION TITLE
Import all Tabs, Folders & Profiles from Arc Browser

### DISCUSSION THREAD

**Discussion by menjilx** [zen-browser/desktop#2770] [Ideas]
### Captchas

- [X] I have read the instructions.
- [X] I have searched existing issues and avoided creating duplicates.
- [ ] I am not filing an enhancement request.

### What happened?

Is adding a feature to import all Arc tabs and folders to Zen and maintaining the folder structures and profiles possible?

TIA

### Reproducible?

- [ ] I have checked that this issue cannot be reproduced on Mozilla Firefox.

### Version

1

### What platform are you seeing the problem on?

macOS - Intel

**Comment by KakouMoses** [comment]
I would ask the same but with the Windows version (I'm currently on Windows 11)

**Comment by rohitt-gupta** [comment]
are there any updates on this feature??

**Comment by mr-cheffy** [comment]
Where does Arc store their bookmarks, etc? Is it the same on macos and windows?

**Reply by udaygirhepunje** [reply]
~/Library/Application Support/Arc/storablesidebar.json

**Comment by pillarcoin** [comment]
Maybe this repo would help.
https://github.com/ivnvxd/arc-export

**Comment by ivanglushko** [comment]
Does Zen have notion of folders like Arc does? It the only feature we lack

**Reply by MaximeCrp** [reply]
It's on its way with #7359

**Comment by Arimodu** [comment]
Ill chip in here to bump this as well. The only thing stopping me and 2 other friends from moving from Arc on Windows to Zen right now is that moving tabs and spaces would be a huge pain in the arse. It is unfortunate but it seems that Arc itself is EOL right now, as TBC is working on their new AI powered thingy, and Arc on windows in its current form is still quite buggy, so there is plenty of reason to switch to another browser.
I am up to implement it myself and PR it in if someone is willing to point me to the import logic location in the codebase, I looked around and couldnt find it, and honestly if I would be spending 2 hours looking through the codebase just to find it I might as well just move my tabs manually.

**Comment by Eatham532** [comment]
Hey, I did some digging for implementing this. I was thinking about doing a PR myself but i've got big exams this year and so don't have enough time to complete it. Instead I decided to dump all of the things I have currently uncovered.

The  `/browser/components/migrate` folder contains the `MigrationUtils.sys.mjs` file that contains links to migrating different browsers and also contains the function that is run by firefox's "import from another browser" wizard. 


Arc uses the `StorableSidebar.json` file for storing all of the tab data. [This repo](https://github.com/ivnvxd/arc-export) contains more information for finding that.


_The rest of this information I found on windows. But I assume that the folder structure will be similar for both the mac and windows apps._  
The User Data is in `[ArcData]\LocalCache\Local\Arc\User Data` (Where `ArcData` is the root path for Arc's app data. Same place as the `StroableSidebar.json` file)

Here you will find a `Default` folder and other folders. Some other folders might be for a profile as well. E.g. I have a `Profile 1` folder in the same `User Data` directory. I suspect if someone had even more profiles, they would be stored here too. 
Inside the `Default` and `Profile 1` folder I found multiple Sqlite databases containing the user data. (History, Login data, Web data, etc...). 


Apologies if the way I organized my findings is confusing. Feel free to ask anything else from me.

**Reply by pickaxe828** [reply]
> The `/browser/components/migrate` folder contains the `MigrationUtils.sys.mjs` file that contains links to migrating different browsers and also contains the function that is run by firefox's "import from another browser" wizard.

Code found in here: https://github.com/mozilla-firefox/firefox/blob/main/browser/components/migration/MigrationUtils.sys.mjs

**Comment by cristoslc** [comment]
+1 -- would give Zen a try in a heartbeat if I could one-click my Arc profiles and spaces. I have multiple profiles (Pesonal, Work, etc.) and multiple spaces within profiles (e.g., one for each active work project) that I'd want to repro in Zen. But with 5 profiles and ~25 spaces, it's too much of a time investment to do manually just to see if I'm as productive in Zen as I am in Arc.

**Reply by cristoslc** [reply]
I haven't tried them yet, but some potentially relevant search resutls:
- https://gist.github.com/clouedoc/4acc8355782f394152d8ce19ceaf53ba | migrating via shuffling Arc profiles to Chrome and using existing import
- https://github.com/ivnvxd/arc-export | converting Arc pinned tabs to bookmarks files (but doesn't help with rest of space config)

**Comment by SamuelMereau** [comment]
Would love to see an update to this...
An integrated solution within Zen would be ideal.

**Comment by jeremycastanza** [comment]
Was considering contributing to this as well... I've been looking at Arc. A full migration would be a little complex to some extent since there are a lot of SQL Lite databases that store different things.

My biggest thing are pinned tabs with folders.  I think once folders are fully stable in the latest release this feature will probably be easy to build. Most of the data looks to be in `StorableSidebar.json`. Looks like the data would just need to be simplified in the JSON file and prepped for some type of import and/or mapping to Zen's structure.

@mr-cheffy - Did you have any additional thoughts on this one? Also, where are these settings going to be in Zen as it relates to folders and such?

**Reply by mr-cheffy** [reply]
My idea was to use Firefox's already existing bookmarks importer and just patch it to make folders instead of bookmark folders

**Reply by jeremycastanza** [reply]
@mr-cheffy - I have a stripped down schema derived from Arc's JSON file. If you think this would be useful for setting the stage for an import or mapping to Zen's folders, LMK the best way to share. I also have a script for generating it from the `StorableSidebar.json`.

**Comment by antifuchs** [comment]
In terms of tools that do the export, I've just used https://xiaogliu.github.io/export-arc-bookmarks/, which (in contrast to the arc-export python tool) runs entirely locally in the browser and manages to get the order of bookmarks and folders right. Can recommend.

Now if only the import step was slightly less convoluted... (:

**Comment by flawnn** [comment]
For less technical people, I have built [Arc Archiver](https://github.com/flawnn/ArcFolderArchiver) as a way to conveniently export whole folders (and soon spaces) as JSON. 

I would love to also extend this to help out with moving away if there is anything that makes sense. The least that I can be helpful with, is providing the scraping & modelling logic for extracting folder data.

I don't use Zen (yet) but was curious what all of you think.

**Reply by hadifarnoud** [reply]
that is a good start. I'd say half way done. 
cheers

**Comment by AntoineHPadoa** [comment]
Hello ! I find a (really) hacky way to import arc folder to zen, and I think it can be generalized to import entire workspaces.
By using internal zen api in browser console, I was able to create folders and populate them with tab (using window.gZenFolders.createFolder, gBrowser.addTab and gBrowser.pinTab). So I write a tiny python scripts which convert arc folders exported as JSON (thanks to @flawnn 's arc archiver) into a js file creating the folders.
I think one could create an entire workspace by digging more zen internal api.
If there is interest, I can make a gist documenting it :)

**Reply by flawnn** [reply]
I added also spaces support now, maybe that helps!

**Reply by antifuchs** [reply]
Do you have a writeup / code for how you did that? I thought `gZenFolders` and others needed to run in the "browser chrome" context, which requires stuff like [fx-autoconfig](https://github.com/MrOtherGuy/fx-autoconfig)?

**Reply by antifuchs** [reply]
Update, my lack of knowledge of mozilla internals was showing: You can access these fields in the browser console (Tools -> Browser Tools -> Browser Console) after setting `devtools.chrome.enabled`.

**Comment by illegalnumbers** [comment]
One thing I noticed is that if you 'Share' a list of open tabs you can save that to a text file. All of the tabs are just line by line. For something massive, like myself (I have over 160 tabs in one workspace...because of course I think I can get through all of that), it would be useful to be able to import just that list if there isn't a way to automate the export of current tabs. `arc-export` does a good job in exporting the spaces and their bookmarks into folders but then you have to import the open tabs after.

**Reply by illegalnumbers** [reply]
Is there a way to get the active tab list from Arc? Does anyone know?

**Reply by illegalnumbers** [reply]
This extension https://addons.mozilla.org/en-US/firefox/addon/tabs-import/?utm_source=addons.mozilla.org&utm_medium=referral&utm_content=search does a pretty good job actually at exactly this thing.

**Reply by illegalnumbers** [reply]
Also if you have the 'New Tab Suspender' extension and set it to 5 seconds that really reduces the memory problem that pops up when you import...hundreds...of tabs.
